### PR TITLE
Multivariate restructure

### PR DESCRIPTION
**Pull request by williamjameshandley**
# Description

This represents a major breaking refactor and extension to the existing suite.

## Before #34

Prior to this PR, we have three major statistical classes:

- stats.multivariate_normal
    - A direct extension to scipy.stats which delivers:
        - conditioning 
        - marginalization 
        - bijector
        - prediction (i.e. rescaling and shifting)
- stats.multimultivariate_normal
    - A non-trivial re-implementation of scipy.stats.multivariate_normal which allows for a 'vector of multivariate normals' with corresponding densities & random variates implemented in addition to the conditioning, marginalization, bijector, and prediction above.
- stats.mixture_multivariate_normal
    - an implementation of a mixture distribution of multivariate normals, with the same functionality as the first, but a structure reminiscent of the second (given this is a vector of normal distributions, but log-sum-exping over the vector.

There are also the corresponding wrapper model classes:

- model.LinearModel
- model.MultiLinearModel
- model.LinearMixtureModel

which make use of the underlying statistical functionality to implement a Bayesian model with likelihood, prior, posterior and evidence.

## In #34

In this PR we make the following changes:

- multimultivariate_normal and multivariate_normal are merged into a single class
    - this is therefore independent of scipy.stats.
- multivariate_normal can now be arbitrary shape (i.e. can be a vector, matrix, or higher-order tensor of normals). 
    - the tensor structure is useful if you had e.g. axes for marginalisating over several parameter values (so that theta can be a tensor argument to likelihood).
- we now have the option to encode and exploit diagonal covariance matrices, which leads to faster inverses, and reduced memory usage.
- mixture_multivariate_normal is now a direct extension of multivariate_normal which marginalises over its final axis by the logA amplitudes.
- There is now a `__getitem__` method for stats, so you can slice the distribution over any of its axes
- The test suite is now heavily extended to cover a much wider variety of cases (crucial for testing something with complicated corner cases).

The same changes have also been applied to LinearModel and MultiLinearModel.

## After #34 (to-do)

The test suite is now quite heavy-duty, and will be subsequently sped up using [this advice](https://github.com/zupo/awesome-pytest-speedup).

The ReducedLinearModel and ReducedLinearModelUniformPrior also needs extension along the above lines, and should also include a long-desired ReducedMixtureModel.

We should offer GPU-able 'torch' options (or equivalent) to this new independent implementation of statistical functions.

# Checklist:

- [X] I have performed a self-review of my own code
- [X] My code is black compliant (`black . --check`)
- [X] My code is isort compliant (`isort . --profile black --filter-files`)
- [X] My code contains compliant docstrings (`pydocstyle --convention=numpy lsbi`)
- [X] New and existing unit tests pass locally with my changes (`python -m pytest`)
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] I have appropriately incremented the [semantic version number](https://semver.org/) in both README.rst and lsbi/_version.py

### CODE REVIEW

**Comment by williamjameshandley** [comment]
@yallup as co-author do you have any views on the API at this stage (pre 1.0.0 nothing is binding).

File: lsbi/model.py [hunk 1]
```diff
@@ -1,236 +1,391 @@
...
+        if ndim>=1: log mixture weights
+        if scalar: scalar * unit vector
+        Defaults to uniform weights
+    n : int, optional
+        Number of parameters, defaults to automatically inferred value
+    d : int, optional
+        Number of data dimensions, defaults to automatically inferred value
+    """
+
+    def __init__(self, logA=1, *args):
```
Comment (yallup) [inline lsbi/model.py]: Two tweaks:

*args, **kwargs

and

logA=1.0

Comment (williamjameshandley) [reply lsbi/model.py]: Many thanks -- ideally there would be a test for this (which hopefully doesn't expand the test suite too much), but before I do that, do you have any opinions on the API in general.
- Should LinearModel have positional arguments, or just kwargs? (the latter is easier to maintain)
- Are we happy with the notation? M,m,mu,C are short, to the point, and relatively standard, Sigma is a bit clunky, but S may not be standard enough. [pep8](https://peps.python.org/pep-3131/#policy-specification) has opinions on Σ and μ, but we could always allow users to provide Sigma and mu as kwargs (it becomes considerably easier with copilot to put in the greek!)

Comment (williamjameshandley) [reply lsbi/model.py]: How cursed does 31924fe look? (certainly benefits from the shorter variable names, but it would be a nightmare for users).

Comment (williamjameshandley) [reply lsbi/model.py]: How cursed does 31924fe look? (certainly benefits from the shorter variable names, but it would be a nightmare for users).

**Comment by yallup** [comment]
Model.model(theta) as a useful function to have

File: lsbi/stats.py [hunk 2]
```diff
@@ -1,165 +1,194 @@
...
-    def bijector(self, x, inverse=False):
-        """Bijector between U([0, 1])^d and the distribution.
-
-        - x in [0, 1]^d is the hypercube space.
-        - theta in R^d is the physical space.
+        return np.exp(self.logpdf(x, broadcast=broadcast))
 
-        Computes the transformation from x to theta or theta to x depending on
-        the value of inverse.
+    def rvs(self, size=()):
```
Comment (yallup) [inline lsbi/stats.py]: lsbi multivariate normal differs from scipy in not squeezing for single rvs

```python
>>> scipy.stats.multivariate_normal(np.ones(100)).rvs(1).shape
(100,)
>>> lsbi.stats.multivariate_normal(np.ones(100)).rvs(1).shape
(1, 100)
```

Comment (williamjameshandley) [reply lsbi/stats.py]: This was a conscious choice, mostly because it makes the code & tests considerably easier to write systematically. Is it a dealbreaker?

Comment (yallup) [reply lsbi/stats.py]: Nope, no issue from my PoV just thought I would flag it as a potential divergence from base structure

Comment (williamjameshandley) [reply lsbi/stats.py]: The default equivalence of rvs(1) and rvs() is arguably a little perverse
```python
>>> scipy.stats.multivariate_normal(np.ones(100)).rvs().shape
(100,)
>>> lsbi.stats.multivariate_normal(np.ones(100)).rvs().shape
(100,)
```

**Comment by yallup** [comment]
> Model.model(theta) as a useful function to have

Now there is a model function, would a ppd wrapper around m.model(m.posterior(data).rvs()) be useful or does that make the namespace too cluttered. If model is a function of theta I suppose ppd() would be a random variable, complement to evidence()

**Comment by yallup** [comment]
Checking I understand the broadcasting rules correctly here

```python
>>> data = model.evidence().rvs(10)
>>> logls = model.likelihood(model.posterior(data).rvs()).logpdf(data)
>>> logls.shape
(10,10)
```

I interpret that in most cases I want the $L(D_i | D_i)$ so I take the diagonal of the matrix above. This seems logical to me, however if the data vector gets large and I am including additional samples ( a new axis) of the conditioned data, this can get heavyweight quickly with a lot thrown away. This may be a very niche application but defaulting to the outer product vector-vector operations may be a bit sticky.

**Comment by williamjameshandley** [comment]
This is not that niche, and is in fact what I've been fixing in the last push so that it works for both multivariates and mixtures.

The fix is to use the `broadcast=True` keyword argument, which are available for both `logpdf` and `rvs`.

```python
>>> logls = model.likelihood(model.posterior(data).rvs()).logpdf(data, broadcast=True)
>>> logls.shape
(10,)
```

Any feedback on the argument name/docstring/default api would be much appreciated.

**Comment by yallup** [comment]
Fairly inconsequential difference but might be an observation for correct typing:

```python
dims = 5
data_dims = dims
np.random.seed(1)

mixtures = 5

Model = LinearModel(
    d=data_dims,
    n=dims,
    C= np.eye(data_dims) * 0.1,
)

MixModel = MixtureModel(
    M = np.random.randn(mixtures, data_dims, dims),
    d = data_dims,
    n = dims,
    C=np.eye(data_dims) * 0.1,
)
```

Processes correctly if you accidentally pass a float instead of int to rvs in one case but not the other, not a case that should work in scipy either so wouldn't expect either case to work

```
Model.evidence().rvs(1.)
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/yallup/projects/fusions/env/lib/python3.11/site-packages/lsbi/stats.py", line 139, in rvs
    x = np.random.randn(*size, *self.shape, self.dim)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "mtrand.pyx", line 1270, in numpy.random.mtrand.RandomState.randn
  File "mtrand.pyx", line 1431, in numpy.random.mtrand.RandomState.standard_normal
  File "_common.pyx", line 636, in numpy.random._common.cont
TypeError: 'numpy.float64' object cannot be interpreted as an integer
MixModel.evidence().rvs(1.)
array([[-2.27897752,  1.37749761, -2.49190091,  4.68398189,  1.22186337]])
```

**Comment by williamjameshandley** [comment]
Good spot. An early unnecessary type conversion was causing the unexpected performance -- fixed now.

**Comment by williamjameshandley** [comment]
FYI, the code will be more efficient using `C=0.1`, or `C=np.ones(data_dims)*0.1`, or `C=np.ones(*any_shape, data_dims)*0.1, diagonal_C=True`, as this avoids the need to invert many data-dimensional matrices

**Comment by yallup** [comment]
```python
model = LinearModel(n=2,d=5)
data = model.evidence().rvs(20)
data.shape
(20, 5)
model.posterior(data).rvs(30,broadcast=True)
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/yallup/projects/lsbi/lsbi/stats.py", line 146, in rvs
    return self.mean + np.sqrt(self.cov) * x
                       ~~~~~~~~~~~~~~~~~~^~~
ValueError: operands could not be broadcast together with shapes (20,2) (30,2) 
```

```python
model = LinearModel(n=2,d=5)
data = model.evidence().rvs(30)
data.shape
(30, 5)
model.posterior(data).rvs(30,broadcast=True)
```

Second one doesnt error (this is a bit of an accidental me getting things the wrong way round so not a real case), but I'm not sure what I actually get out of this broadcast rule in this case? is it equivalent to `[model.posterior(di).rvs() for di in data]`

**Comment by williamjameshandley** [comment]
The broadcast for rvs is a bit of a niche case (mostly useful for the [dkl calculation](https://github.com/handley-lab/lsbi/blob/6382422cb7567fc02fef73132811164246867a91/lsbi/model.py#L461-L464).

Here is example usage:
```python
>>> dist.shape
(10)
>>> dist.rvs().shape
(10,5)
>>>  dist.rvs(10, broadcast=True).shape # equivalent to the previous
(10,5)
>>> dist.rvs(20).shape
(20,10,5)
>>> dist.rvs((20,10), broadcast=True).shape # equivalent to the previous
(20,10,5)
```
i.e. it is really away of 'ignoring' the shape of the distribution, leaving it up to the user to broadcast correctly.

How could the docstrings be clearer?

**Review by yallup — APPROVED** [review]
lgtm

**Comment by williamjameshandley** [comment]
Many thanks @yallup. I still read lgtm as let's get this merged.
